python/ray/serve/_private/replica.py

```diff
@@ -814,145 +811,199 @@ async def call_reconfigure(self, user_config: Any):
-                    "You must use `handle.options(stream=True)` to call "
-                    "generators on a deployment."
-                )
+        Returns (request_args, request_kwargs).
+        """
+        request_args = (pickle.loads(request.grpc_user_request),)
+        if GRPC_CONTEXT_ARG_NAME in user_method_params:
+            request_kwargs = {GRPC_CONTEXT_ARG_NAME: request_metadata.grpc_context}
+        else:
+            request_kwargs = {}
 
-            method_to_call = sync_to_async(runner_method)
+        return request_args, request_kwargs
 
-            # Edge case to support empty HTTP handlers: don't pass the Request
-            # argument if the callable has no parameters.
-            params = inspect.signature(runner_method).parameters
-            if request_metadata.is_http_request and len(params) == 0:
-                request_args, request_kwargs = tuple(), {}
-            elif request_metadata.is_grpc_request and GRPC_CONTEXT_ARG_NAME in params:
-                request_kwargs = {GRPC_CONTEXT_ARG_NAME: request_metadata.grpc_context}
-
-            result = await method_to_call(*request_args, **request_kwargs)
-            if inspect.isgenerator(result) or inspect.isasyncgen(result):
+    async def _handle_user_method_result(
+        self,
+        result: Any,
+        user_method_name: str,
+        request_metadata: RequestMetadata,
+        *,
+        result_queue: Optional[MessageQueue],
+        is_asgi_app: bool,
+        asgi_args: Optional[ASGIArgs],
+    ) -> Any:
+        """Postprocess the result of a user method.
+
+        If this is for a streaming request, the result should be a generator that will
+        be consumed and its outputs placed on the `result_queue`.
+
+        This method may raise an exception if the result is not of the expected type
+        (e.g., non-generator for streaming requests or generator for unary requests).
+        """
+        result_is_gen = inspect.isgenerator(result)
+        result_is_async_gen = inspect.isasyncgen(result)
+        if request_metadata.is_streaming:
+            if result_is_gen:
+                for r in result:
+                    if request_metadata.is_grpc_request:
+                        r = (request_metadata.grpc_context, r.SerializeToString())
+                    await result_queue(r)
+            elif result_is_async_gen:
+                async for r in result:
+                    if request_metadata.is_grpc_request:
+                        r = (request_metadata.grpc_context, r.SerializeToString())
+                    await result_queue(r)
+            elif request_metadata.is_http_request and not is_asgi_app:
+                # For the FastAPI codepath, the response has already been sent over
+                # ASGI, but for the vanilla deployment codepath we need to send it.
+                await self._send_user_result_over_asgi(result, asgi_args)
+            elif not request_metadata.is_http_request:
```

Comment:
Can you help me understand why this one is elif rather than just else?

Reply:
ah yes this is indeed a little bit confusing... for HTTP handlers, the user-provided method will actually be unary but we are still streaming back results via the result queue. This is because we follow the `ASGI` protocol, so the method looks like this:

```python
async def __call__(self, scope, receive, send):
    # use `send` to ship back ASGI messages. In our case, `send` is the result queue.
```

Let me add a comment to this effect in the code since it's certainly not obvious.

Reply:
(added)

Reply:
Don't we still need to have an else block here? 

Otherwise, if i'm calling gRPC unary, but specify `stream=True` we seem to just drop it on the floor and won't report any error

Reply:
That'll fall into the `elif not request_metadata.is_http_request` block

Reply:
This block is essentially whitelisting that it's ok to specify `stream=True` but have a unary method _only_ for HTTP requests (all others will raise the error)

Reply:
Ok, i see what you meant now. Thanks for updating the comment it's more clear now!